src/requests/create-signature.ts

```diff
@@ -0,0 +1,108 @@
+import * as crypto from 'crypto'
+import type { CanonicalRequest, Secret, Timestamp } from './typings'
+import { CanonicalRequestValidator, SecretValidator, TimestampValidator } from './typings'
+import { getNormalizedEncodedURI, getNormalizedHeaders } from './utils'
+import { ContentfulSigningHeader } from './constants'
+
+const hash = (normalizedCanonicalRequest: CanonicalRequest, secret: string) => {
+  const stringifiedHeaders = Object.entries(normalizedCanonicalRequest.headers!)
+    .map(([key, value]) => `${key}:${value}`)
+    .join(';')
+
+  const stringifiedRequest = [
+    normalizedCanonicalRequest.method,
+    normalizedCanonicalRequest.path,
+    stringifiedHeaders,
+    normalizedCanonicalRequest.body,
+  ].join('\n')
+
+  const hmac = crypto.createHmac('sha256', secret)
+
+  hmac.update(stringifiedRequest)
+
+  return hmac.digest('hex')
+}
+
+const enrichHeadersWithMetadata = (
+  headers: Record<string, string>,
+  signedHeaders: string[],
+  timestamp: number
+) => {
+  const result: Record<string, string> = {}
+  const joinedSignedHeaders = signedHeaders.join(',')
+
+  result[ContentfulSigningHeader.Timestamp] = timestamp.toString()
+
+  if (joinedSignedHeaders) {
+    result[ContentfulSigningHeader.SignedHeaders] = joinedSignedHeaders
+  }
+
+  return { ...headers, ...result }
+}
+
+/**
+ * Given a secret, a canonical request and a timestamp, generates a signature.
+ * It can be used to verify canonical requests to assess authenticity of the
+ * sender and integrity of the payload.
+ *
+ * ~~~
+ * const {createSignature} = require('contentful-node-apps-toolkit')
+ * const {server} = require('./imaginary-server')
+ *
+ * const SECRET = process.env.SECRET
+ *
+ * server.post('/api/my-resources', (req, res) => {
+ *   const incomingSignature = req.headers['x-contentful-signature']
+ *   const incomingTimestamp = req.headers['x-contentful-timestamp']
+ *   const now = Date.now()
+ *
+ *   if (!incomingSignature) {
+ *     res.send(400, 'Missing signature')
+ *   }
+ *
+ *   if (now - incomingTimestamp > 1000) {
+ *     res.send(408, 'Request too old')
+ *   }
+ *
+ *   const computedSignature = createSignature(
+ *     SECRET,
+ *     {
+ *       method: req.method,
+ *       path: req.url,
+ *       signedHeaders: ['Authorization']
```

Comment:
Thinking about it, I wouldn't do `signedHeaders` option. I would sign only headers passed and include computed `x-contentful-signed-headers` for signature. When verifying, I would read the header value and pick selected headers from the hash of all headers, pass it and compare. We should hide it in `verify()` utility

Reply:
The problem I see with that is sorting (which is the reason why first implementation was exactly as you suggested + sorting): objects are not guaranteed to have sorted keys, so we would need a data structure which enforces order (as array, comma-separated strings...) in order to have that consistency. Shall we go back to alphabetical order or do we have different suggestions?

Reply:
I think we should just always sort keys in the object provided